File: sliding_max.rb
require_relative "stack"
# Array version
def do_stuff(arr, k, output = "")
  return arr.max.to_s if arr.size <= k
  arr2 = arr.slice(0,k)
  output += arr2.max.to_s + " " + do_stuff(arr.slice(1..-1), k, output)
end

def do_stuff_list(arr, k, output = "")
  stack = Stack.new
  arr.slice(0...k).each {|el| stack.push(el)}
  nxt = arr.slice(1..-1).size < k ? "" : do_stuff_list(arr.slice(1..-1),k,output)
  output += stack.max.to_s + " " + nxt
end

k, input, input2, input3 = [3, [1, 3, 5, 7, 9, 2],[9, 7, 3, 0, -1, 11, 3], [2, 2, 3, 3, 7, 7, 2, 2, 3, 3]
]

puts do_stuff(input, k)
puts do_stuff(input2, k)
puts do_stuff(input3, k)

puts do_stuff_list(input,k)
puts do_stuff_list(input2,k)
puts do_stuff_list(input3,k)

File: slidingMax.js
require "./stack.js"
// Array version
function do_stuff(arr, k, output = "") {
  if (arr.length === k) return Math.max(...arr)
  const arr2 = arr.slice(0, k);
  return output += Math.max(...arr2) + " " + do_stuff(arr.slice(1), k, output);
}

// Linked List version
function do_stuff_list(arr, k, output = "") {
  const stack = new Stack;
  arr.slice(0, k).forEach(v => stack.push(v));
  const next = arr.slice(1).length < k ? "" : do_stuff_list(arr.slice(1), k, output)
  return output += stack.max() + " " + next;
}

const [k, input] = [3, [1, 3, 5, 7, 9, 2]]